### PR TITLE
Feature/web api internal traps

### PR DESCRIPTION
**Pull request by louismunro**
For your consideration. 
This adds support for deauthentication through the WebAPI.

Not yet called by default. We still have to modify enforcement.pm for that.

### CODE REVIEW

File: lib/pf/api.pm [hunk 1]
```diff
@@ -182,6 +182,91 @@ sub firewallsso {
     return $pf::config::TRUE;
 }
 
+
+sub ReAssignVlan {
+    my ($class, %postdata )  = @_;
+    my $logger = Log::Log4perl->get_logger('pf::WebAPI');
```
Comment (cgx) [inline lib/pf/api.pm]: Simply use `get_logger` instead (defined in `pf::log`).

File: NEWS.asciidoc [hunk 2]
```diff
@@ -45,6 +45,7 @@ Enhancements
 * On status page check the portal associated to the user and authenticate on the sources included in the portal profile
 * Merge pf::email_activation and pf::sms_activation to pf::activation
 * Removed unused table switchlocation
+* Deauthentication and firewall enforcement can now be done throught the web API.
```
Comment (extrafu) [inline NEWS.asciidoc]: remove trailing dot

File: lib/pf/api.pm [hunk 3]
```diff
@@ -182,6 +182,91 @@ sub firewallsso {
...
+        return;
+    }
+
+    my $switch = pf::SwitchFactory->getInstance()->instantiate( $postdata{'switch'} );
+
+    # SNMP traps connections need to be handled specially to account for port-security etc.
+    if ( ($postdata{'connection_type'} & $WIRED_SNMP_TRAPS) == $WIRED_SNMP_TRAPS ) {
+        _reassignSNMPConnections($switch, $postdata{'mac'}, $postdata{'ifIndex'}, $postdata{'connection_type'} );
+    }
+    elsif (  $postdata{'connection_type'} & $WIRED ) {
```
Comment (extrafu) [inline lib/pf/api.pm]: One too many space after (

File: lib/pf/api.pm [hunk 4]
```diff
@@ -182,6 +182,91 @@ sub firewallsso {
...
+
+sub desAssociate {
+    my ($class, %postdata )  = @_;
+    my $logger = Log::Log4perl->get_logger('pf::WebAPI');
+
+    my $switch = pf::SwitchFactory->getInstance()->instantiate($postdata{'switch'});
+
+    my ($switchdeauthMethod, $deauthTechniques) = $switch->deauthTechniques($switch->{'_deauthMethod'});
+
+    $logger->info("DeAssociating mac $postdata{'mac'} on switch " . $switch->{_id});
```
Comment (extrafu) [inline lib/pf/api.pm]: desAssociate vs. deAssoci.. ?

File: lib/pf/api.pm [hunk 5]
```diff
@@ -182,6 +182,91 @@ sub firewallsso {
...
+
+    if ( not defined( $postdata{'connection_type'} )) { 
+        $logger->error("Connection type is unknown. Could not reassign VLAN."); 
+        return;
+    }
+
+    my $switch = pf::SwitchFactory->getInstance()->instantiate( $postdata{'switch'} );
+
+    # SNMP traps connections need to be handled specially to account for port-security etc.
+    if ( ($postdata{'connection_type'} & $WIRED_SNMP_TRAPS) == $WIRED_SNMP_TRAPS ) {
```
Comment (fdurand) [inline lib/pf/api.pm]: $pf::config::WIRED_SNMP_TRAPS instead

File: lib/pf/api.pm (hunk 3 above)
Comment (fdurand) [inline lib/pf/api.pm]: $pf::config::WIRED instead

File: lib/pf/api.pm (hunk 4 above)
Comment (louismunro) [inline lib/pf/api.pm]: Technically deAssociate is the correct form.
desAssociate is a gallicism.

Comment (louismunro) [inline lib/pf/api.pm]: Should we stick to desAssociate for consistency though?

Comment (extrafu) [inline lib/pf/api.pm]: Indeed but my comment was related on using the same formulation everywhere :-)

Comment (louismunro) [inline lib/pf/api.pm]: Ok. Consistency wins. Will update accordingly.

File: lib/pf/api.pm [hunk 6]
```diff
@@ -182,6 +182,91 @@ sub firewallsso {
...
+
+    if ( not defined( $postdata{'connection_type'} )) { 
+        $logger->error("Connection type is unknown. Could not reassign VLAN."); 
+        return;
+    }
+
+    my $switch = pf::SwitchFactory->getInstance()->instantiate( $postdata{'switch'} );
+
+    # SNMP traps connections need to be handled specially to account for port-security etc.
+    if ( ($postdata{'connection_type'} & $WIRED_SNMP_TRAPS) == $pf::config::WIRED_SNMP_TRAPS ) {
```
Comment (fdurand) [inline lib/pf/api.pm]: Still $WIRED_SNMP_TRAPS

**Pull request merged**